src/device/usbd.c

```diff
@@ -758,9 +771,13 @@ static bool process_get_descriptor(uint8_t rhport, tusb_control_request_t const
       uint16_t len = sizeof(tusb_desc_device_t);
 
       // Only send up to EP0 Packet Size if not addressed
+      // This only happens with the very first get device descriptor and EP0 size = 8 or 16.
       if ((CFG_TUD_ENDPOINT0_SIZE < sizeof(tusb_desc_device_t)) && !_usbd_dev.addressed)
       {
         len = CFG_TUD_ENDPOINT0_SIZE;
+
+        // Hack here: we modify the request length to prevent usbd_control response with zlp

```

Comment:
@pigrew @cr1901 this is a hack to modify wLength, else usbd_control will send out 8 byte + ZLP. This is one time hack and only needed for msp430 (and probably 8-bit mcu), I don't know anyway cleaner.